### PR TITLE
Feat: Compute test grades

### PR DESCRIPTION
**Pull request by carissa-tang**
## Notion ticket link

[Compute test grades](https://www.notion.so/uwblueprintexecs/Compute-test-grades-4c7b0c281b7b4e02be33b55bd500a5c5)

## Implementation description
* computeTestGrades takes in a `UngradedResultDTO` and its associated test id, and returns a `GradedResultResponseDTO` with the calculated score and breakdown
* Will be called within `updateTestSession` > `createTestResult`
   * Can look here https://github.com/uwblueprint/jump-math/pull/36/files for full picture of how this function intends to be used
* Wrote tests
   * Successful case
   * Invalid test id 

## Steps to test
1.

## What should reviewers focus on?
* 

## Checklist
- [ ] My PR name is descriptive and in imperative tense
- [ ] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [ ] I have run the appropriate linter(s)
- [ ] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/services/implementations/testSessionService.ts
Comment (carissa-tang) [inline backend/services/implementations/testSessionService.ts]: I don't know if this will even be a possible case, but I have it here anyways

Side note: Is it possible for a student to leave a question blank? If so, should we check if the answer=null or smth @joyce-shi

Comment (joyce-shi) [reply backend/services/implementations/testSessionService.ts]: Hmm,  I don't think so. If a student left the answer blank, I think best practice would be to populate from the frontend with a `null` value. So, given this, I'm not sure if `test.questions.length` would ever vary from the `studentAnswers.length`? Let me know if I'm missing a case here!

**Comment by joyce-shi** [comment]
Side Note: interesting how `yarn fix` seemed to make changes... maybe a good example for why we should have automatic pre-commit linting... I feel like those sort of changes make it harder to review PRs

File: backend/services/interfaces/testSessionService.ts
Comment (joyce-shi) [inline backend/services/interfaces/testSessionService.ts]: Hmmm... so I think we had some context for why we had kept Response and Request DTOs separate (despite their similarities), see [here](https://github.com/uwblueprint/jump-math/pull/25#discussion_r917479900). I think Mahad makes a good point about keeping them separate for the sake of modularity!

If you're interested, you can read a bit more about this [here](https://stackoverflow.com/questions/44349199/reusing-dto-for-various-request-response-types-vs-explicitness-of-what-is-requir)! (Although, the link talks about separating common elements into some sort of base class, but I think that's overkill for our use case here)

File: backend/services/interfaces/testSessionService.ts
Comment (joyce-shi) [inline backend/services/interfaces/testSessionService.ts]: What are your thoughts on calling this something like `UnscoredResultDTO` or `UnmarkedResultDTO` or something? To include a bit more context in naming & make it a bit more intuitive/descriptive. Or something else if you can think of a better name than those LOL.

Comment (joyce-shi) [reply backend/services/interfaces/testSessionService.ts]: Let's also add some documentation for the interface, explaining why its necessary as well (i.e what we talked about on call last work session).

File: backend/services/implementations/testSessionService.ts
Comment (joyce-shi) [inline backend/services/implementations/testSessionService.ts]: Re: similar to the comment made in `interfaces/testSessionService.ts` let's call this `markedResult` or `gradedResult` or something along those lines.

File: backend/services/implementations/testSessionService.ts
Comment (joyce-shi) [inline backend/services/implementations/testSessionService.ts]: baaaaa... I think when we defined models, we said that this would always be of type `number`, and if it was short answer it would be the numeric answer and if it was multiple choice, it would be the index. What are thoughts on adding a quick comment here that this is the case? Because I think this logic is not super intuitive.

And also updating the `testSession.model.ts` to say something to that effect as well!

File: backend/services/implementations/testSessionService.ts
Comment (joyce-shi) [inline backend/services/implementations/testSessionService.ts]: Not sure if this check is necessary, as discussed [here](https://github.com/uwblueprint/jump-math/pull/40#discussion_r932817119). Let me know if I'm missing something!

File: backend/services/implementations/testSessionService.ts
Comment (joyce-shi) [inline backend/services/implementations/testSessionService.ts]: nit: let's call this `computedScore` or something, since I don't think we have a score before, so don't think we are updating anything?

File: backend/services/implementations/testSessionService.ts
Comment (joyce-shi) [inline backend/services/implementations/testSessionService.ts]: same as above ^^.

File: backend/services/implementations/testSessionService.ts
Comment (carissa-tang) [reply backend/services/implementations/testSessionService.ts]: Wait so, on backend would we store smth like `studentAnswers=[10, null, 20]` if second question is blank

**Review by joyce-shi — CHANGES_REQUESTED** [review]
Left a couple comments! Mostly around code readability and variable naming. Let me know if you have thoughts please!

Also, general comment about this PR: I noticed that there are a lot of comments throughout the body of the `computeTestGrades` function. While I do think we've been adding a lot of comments in this project in general, I think up until now it's been okay because code comments are generally good practice for function and interface declarations. This is because they allow for future developers to skim through code and see what functions do without spending too much time in the weeds of the code. (More info [here](https://codingart.readthedocs.io/en/latest/c/Comments.html#:~:text=Every%20function%20declaration%20should%20have,comments%20in%20the%20function%20definition.)!)

But in the body of the function, I think general consensus is that there should only be code comments when there is a non-intuitive case (and there is no way to make it intuitive, i.e the fact that we store all `studentAnswers` as numbers). So for things like "updating the counter", there should be no need to add a comment! If it's not clear that we are incrementing the counter, maybe it is helpful to rename the variable, i.e include the term counter, or something.

Some best practices for commenting can be found [here](https://stackoverflow.blog/2021/12/23/best-practices-for-writing-code-comments/)! If you're interested in learning more. I also think that when it comes to this sort of code styling / readability commentary, it gets pretty subjective, so use your best judgement and of course ping me if you want to discuss further / have questions!

(Also, sorry for such a long monologue LOL, I hope this is helpful and interesting!)

But also, 🔥 work overall! I think you've been working on two of the more complex tickets out of the backend work we've been doing and I'm impressed with your clarity and comprehensiveness of thinking around how the code should function. (:

File: backend/services/implementations/testSessionService.ts
Comment (joyce-shi) [reply backend/services/implementations/testSessionService.ts]: I think so. Let me know if you think there is a better alternative! And if we do this, we should have `null` checks in the `computeGrades` function then as well, I think.

File: backend/testUtils/testSession.ts
Comment (carissa-tang) [inline backend/testUtils/testSession.ts:57]: I've updated the result interfaces to `(number | null)[]` to allow us to store missing answers as null - lmk if this makes sense or if there's a better implementation / or if we are even allowing for blank answers in the first place? (can just make it required field in FE)

File: backend/services/implementations/testSessionService.ts
Comment (carissa-tang) [reply backend/services/implementations/testSessionService.ts]: i've updated the interfaces! should work with the current logic since we directly check for strict equality `studentAnswers[i] === actualAnswer`

File: backend/services/interfaces/testSessionService.ts
Comment (mahadzaryab1) [inline backend/services/interfaces/testSessionService.ts]: i'm wondering why we need to separate result objects? can we not just have one with the `breakdown` and `score` set to null or empty?

Comment (mahadzaryab1) [reply backend/services/interfaces/testSessionService.ts]: we could potentially even add a `status` field and have it default to `ungraded` and then set it to `graded` once it is.

File: backend/testUtils/testSession.ts
Comment (mahadzaryab1) [inline backend/testUtils/testSession.ts]: should this go in `testUtils/test.ts`?

File: backend/services/implementations/testSessionService.ts
Comment (mahadzaryab1) [inline backend/services/implementations/testSessionService.ts]: can we switch this to an `else if`? just in case we decide to introduce another question type. Another option is to use a switch statement

File: backend/services/implementations/testSessionService.ts
Comment (mahadzaryab1) [inline backend/services/implementations/testSessionService.ts]: we can move this entire block into a helper function called `getCorrectAnswer` or something

File: backend/testUtils/testSession.ts
Comment (mahadzaryab1) [inline backend/testUtils/testSession.ts]: let's have one of these as a float to ensure it holds there

**Review by mahadzaryab1 — COMMENTED** [review]
nice work - added a few comments

File: backend/models/testSession.model.ts
Comment (joyce-shi) [inline backend/models/testSession.model.ts:22]: beauty ty 🔥

**Review by joyce-shi — APPROVED** [review]
LGTM. I think Mahad has a few comments outstanding, but once you address those you can merge!

File: backend/services/interfaces/testSessionService.ts
Comment (carissa-tang) [reply backend/services/interfaces/testSessionService.ts]: i like this suggestion! @joyce-shi what do you think

File: backend/services/implementations/testSessionService.ts
Comment (carissa-tang) [reply backend/services/implementations/testSessionService.ts]: switch won't work cause `QuestionType.MULTIPLE_CHOICE=0` always evaluates to true :(. i wonder if there's a way to make it work... but will use if statement for simplicity

Comment (mahadzaryab1) [reply backend/services/implementations/testSessionService.ts]: I think it's possible - see [here](https://stackoverflow.com/questions/27747437/typescript-enum-switch-not-working). I'll leave it up to you if you want to change it

**Review by mahadzaryab1 — APPROVED** [review]
lgtm! great work

**Pull request merged**
